File: rope-node.go
package main

import (
	"github.com/koding/kite"
)

func main() {
	r := kite.New("dope", "0.0.0")
	// r.SetLogLevel(kite.DEBUG)
	// r.Config.DisableAuthentication = true

	kiteURL := "ws://localhost:8080"
	l := r.NewClient(kiteURL)

	api := map[string]kite.HandlerFunc{
		"square": func(req *kite.Request) (interface{}, error) {
			number := req.Args.One().MustFloat64()
			result := number * number
			return result, nil
		},
		"identified": func(req *kite.Request) (interface{}, error) {
			res, err := l.Tell("query", nil)
			if err != nil {
				panic(err)
			}

			var kites []string
			if err := res.Unmarshal(&kites); err != nil {
				panic(err)
			}

			r.Log.Info("Following Kites found on Rope: %v\n", kites)
			return nil, nil
		},
	}

	for method, f := range api {
		r.HandleFunc(method, f)
	}

	r.HandleFunc("identify", func(req *kite.Request) (interface{}, error) {
		r.Log.Info("Identify requested!")
		funcs := make([]string, 0, len(api))
		for method := range api {
			funcs = append(funcs, method)
		}
		return map[string]interface{}{
			"kiteInfo": r.Kite(),
			"api":      funcs,
		}, nil
	})

	err := l.Dial()
	if err != nil {
		r.Log.Fatal(err.Error())
	}
	r.Run()
}

File: rope-node.js
var window = window || {}
const Kite = window.Kite || require('kite.js').Kite

const AUTO_RECONNECT = false
const LOG_LEVEL = 4
const ENVIRONMENT = window.Kite ? 'browser-environment' : 'node-environment'
const NAME = 'math'

var publicKites = []

const api = {
  identified: function(id) {
    kite.emit('info', 'Now identified with', id)

    kite.tell('query', [null]).then(function(res) {
      kite.emit('info', 'following kites found on rope', (publicKites = res))
    })
  },
  identify: function(id, callback) {
    // kite.emit('info', 'identify requested, doing now...', id)
    callback(null, { api: api, kiteInfo: kite.getKiteInfo() })
  },
  square: function(number, callback) {
    callback(null, number * number)
  },
}

var kite = new Kite({
  url: 'ws://0.0.0.0:8080/websocket',
  name: NAME,
  logLevel: LOG_LEVEL,
  environment: ENVIRONMENT,
  autoReconnect: AUTO_RECONNECT,
  api: api,
})

// console.log(kite.api.methods)
kite.connect()
